common/app/routes/Settings/components/Cert-Settings.jsx

```diff
@@ -102,23 +120,154 @@ class CertificationSettings extends PureComponent {
     }
   }
 
-  handleSubmit(values) {
-    const { id } = values;
-    const fullForm = _.values(values)
+  buildProjectForms({
+    projectBlockName,
+    challenges,
+    superBlock
+  }) {
+    const {
+      blockNameIsCertMap,
+      claimCert,
+      hardGoTo,
+      userProjects,
+      username
+    } = this.props;
+    const isCertClaimed = blockNameIsCertMap[projectBlockName];
+    if (superBlock === jsProjectSuperBlock) {
+      return (
+        <JSAlgoAndDSForm
+          challenges={ challenges }
+          claimCert={ claimCert }
+          hardGoTo={ hardGoTo }
+          isCertClaimed={ isCertClaimed }
+          jsProjects={ userProjects[superBlock] }
+          key={ superBlock }
+          projectBlockName={ projectBlockName }
+          superBlock={ superBlock }
+          username={ username }
+        />
+      );
+    }
+    const options = challenges
+      .reduce((options, current) => {
+        options.types[current] = 'url';
+        return options;
+      }, { types: {} });
+
+    options.types.id = 'hidden';
+    options.placeholder = false;
+
+    const userValues = userProjects[superBlock] || {};
+
+    if (!userValues.id) {
+      userValues.id = superBlock;
+    }
+
+    const initialValues = challenges
+      .reduce((accu, current) => ({
+        ...accu,
+        [current]: ''
+      }), {});
+
+    const completedProjects = _values(userValues)
       .filter(Boolean)
-      .filter(_.isString)
-      // 5 projects + 1 id prop
-      .length === 6;
-    const valuesSaved = _.values(this.props.userProjects[id])
+      .filter(isString)
+      // minus 1 to account for the id
+      .length - 1;
+
+    const fullForm = completedProjects === challenges.length;
+    return (
+      <FullWidthRow key={superBlock}>
+        <h3 className='project-heading'>{ projectBlockName }</h3>
+        <Form
+          buttonText={ fullForm ? 'Claim Certificate' : 'Save Progress' }
+          enableSubmit={ fullForm }
+          formFields={ challenges.concat([ 'id' ]) }
+          hideButton={isCertClaimed}
+          id={ superBlock }
+          initialValues={{
+            ...initialValues,
+            ...userValues
+          }}
+          options={ options }
+          submit={ this.handleSubmit }
+        />
+        {
+          isCertClaimed ?
+            <Button
+              block={ true }
+              bsSize='lg'
+              bsStyle='primary'
+              href={ `/c/${username}/${superBlock}`}
+              >
+              Show Certificate
+            </Button> :
+            null
+        }
+        <hr />
+      </FullWidthRow>
+    );
+  }
+
+  isProjectSectionCompleted(values) {
+    const { id } = values;
+    const { projects } = this.props;
+    const whereSuperBlockIsId = compareSuperBlockWith(id);
+
+    let pIndex = findIndex(projects, whereSuperBlockIsId);
```

Comment:
Instead of using the imperative approach with `findIndex`, why not use `find` to pull out the item you need instead?